Scan/STLcd/lcd_scan.c

```diff
@@ -338,11 +345,204 @@ inline void LCD_setup()
 	PORTC_PCR3 = PORT_PCR_SRE | PORT_PCR_DSE | PORT_PCR_MUX(4);
 }
 
+static uint32_t STLcdSyncStartMillis;
+static uint32_t STLcdSyncStartTicks;
+
+// Display buffer
+uint8_t STLcdBuffer[LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES];
+uint8_t STLcdSyncBuffer[LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES];
+
+
+// Bounding box.
+static uint8_t STLcdUpdateXMin = LCD_WIDTH, STLcdUpdateXMax = 0, STLcdUpdateYMin = LCD_HEIGHT, STLcdUpdateYMax = 0;
+static uint8_t STLcdSync = 0;
+static uint8_t STLcdSyncPageMin, STLcdSyncPageMax, STLcdSyncColumnMin, STLcdSyncColumnMax;
+
+static void STLcd_updateBoundingBox(uint8_t xmin, uint8_t ymin, uint8_t xmax, uint8_t ymax) {
+    if (xmin < STLcdUpdateXMin) STLcdUpdateXMin = xmin;
+    if (xmax > STLcdUpdateXMax) STLcdUpdateXMax = xmax;
+    if (ymin < STLcdUpdateYMin) STLcdUpdateYMin = ymin;
+    if (ymax > STLcdUpdateYMax) STLcdUpdateYMax = ymax;
+}
+
+void STLcd_clear(void) {
+    memset(STLcdBuffer, 0, LCD_PAGE_LEN * LCD_TOTAL_VISIBLE_PAGES);
+    STLcd_updateBoundingBox(0, 0, LCD_WIDTH, LCD_HEIGHT);
+}
+
+static uint8_t STLcdSyncPageCurrent;
+static uint8_t *STLcdSyncBufferColumnCurrent;
+static uint8_t *STLcdSyncBufferColumnMin, *STLcdSyncBufferColumnMax;
+static uint32_t STLcdSyncStage;
+
+#define STAGE_WRITE_START 15
+#define STAGE_WRITE_END 20
+
+void STLcd_blockingSync(void) {
+    for(;STLcdSyncPageCurrent < STLcdSyncPageMax;
+	STLcdSyncPageCurrent++, STLcdSyncBufferColumnMin += LCD_PAGE_LEN)
+    {
+	LCD_writeControlReg(0xB0 | STLcdSyncPageCurrent);
+	LCD_writeControlReg(0x10 | STLcdSyncColumnMin >> 4);
+	LCD_writeControlReg(0x10 | (STLcdSyncColumnMin & 0x0f));
+	SPI_write(STLcdSyncBufferColumnMin, STLcdSyncColumnMax - STLcdSyncColumnMin);
+    }
+}
+
+// a non-blocking sync function
+void STLcd_sync(void) {
```

Comment:
I am a newbie to the embedded system, and these code were written in a hurry(I need them to implement a basic calculator on the keyboard).
To my knowledge. If this method is to be used in other modules. The first thing should be moving all the signal test  and time test  things to a main loop. The main loop will call the function when it is able to continue(a reactor pattern). This can reduce the function call overhead. Maybe there is already a good coroutine library for this. And if there is not. We can still use the pycparser module to write a script that will convert a function to a non-blocking one.